core/src/main/java/io/grpc/internal/TimeProvider.java

```diff
@@ -29,7 +29,10 @@ public interface TimeProvider {
   TimeProvider SYSTEM_TIME_PROVIDER = new TimeProvider() {
     @Override
     public long currentTimeNanos() {
-      return TimeUnit.MILLISECONDS.toNanos(System.currentTimeMillis());
```

Comment:
Revert this change. This code is used by all callers. We may only call java.time when the caller is using java.time, as java.time is not always available. We learn that java.time is available by someone calling us with a java.time object.

Reply:
Can you elaborate how java.time is needed by java.util.concurrent.TimeUnit? The change was done because grpc needs Java 8 and TimeUnit's toNanos is supported in Java 8.

Reply:
Oh, I thought this was for #10245. java.time was added in Android API level 26. https://developer.android.com/reference/java/time/package-summary . We are supporting 21. We run animalsniffer to detect cases like this, but apparently it is broken, because this is the third(?) time recently it hasn't detected a problem.

If we want to use java.time, we'd probably want two separate time providers and detect which to use at runtime.

Reply:
Ok, but this PR uses java.util.concurrent and the fact that Java 8 supports nanoseconds (and Android API level being orthogonal to this ability).

Reply:
What uses `java.util.concurrent`? I see `import java.time.Instant;` above. Shared projects like this (vs something non-Android like xds) use the intersection of {Java 8, Android API 21}, as otherwise it won't work on a supported platform. If we really want something that is only on one platform, then generally we need reflection and a fallback.

Reply:
Got it, I misread the package being used because the comment shows in GitHub UI on the left panel (old code).

Reply:
@vinodhabib we need to avoid having imports for Java.time.Instant. We need to use reflection (such as Class.forName("Java.time.Instant")) to check if the class is available, and if so do the new changes, but **also using reflection**., and if not available, continue to do the old way using java.util.concurrent.TimeUnit and the nanoseconds offset logic that was there before. There won't be any occurrence of the Instant class statically compiled.

Reply:
Calling methods via reflection has a performance overhead (the creation of the Instant object as well as calling the toNanos method both have to be done via reflection everytime TimeProvider.currentTimeNanos is called, we cannot create and keep the Instant object at startup time and do just 1 method call getNanos on this object via reflection everytime the method is called), is it worth to do this change to use Java.time's nanos? Can we leave it as it is?
 @vinodhabib please hold off your changes for now.

Reply:
Leaving it as is is definitely an option.
If trying to implement this - can be 2 implementations of `TimeProvider` and `SYSTEM_TIME_PROVIDER` initialised once at the moment of class loading to one or another.

Reply:
yet to work on this change.

Reply:
Fixed